### PR TITLE
feat: Added deadlines to refinancing and allowed pool delegates to explicitly reject new terms  (SC-4886)

### PR DESCRIPTION
**Pull request by vbidin**
# Description
Added additional functionality to the `MapleLoan` contract which allows all refinance proposals to include deadlines up to which they can be accepted by the pool delegate, and added functionality that allows the borrower or lender to explicitly reject the terms of the refinancing.

# Integrations Checklist

- [x] Have any function signatures changed? If yes, outline below.
- [x] Have any features changed or been added? If yes, outline below.
- [x] Have any events changed or been added? If yes, outline below.
- [x] Has all documentation been updated?

# Changelog

## Features
- Added a `deadline` parameter to the `acceptNewTerms()` and `proposeNewTerms()` functions, and introduced a check that prevents the terms from being accepted when calling `acceptNewTerms()` after the deadline has elapsed.
- Added an additional function called `rejectNewTerms()` that allows pool delegates to explicitly reject terms of the proposed refinancing.
- Added an event called `NewTermsRejected` which is emitted when anyone explicitly rejects the proposed terms by calling `rejectNewTerms()`.
- Updated existing tests and added additional unit tests which test out scenarios where deadlines are used and proposals are rejected.

## Function Signature Changes
Updated the following functions to include the `deadline` parameter in their signature:
- `acceptNewTerms()`
- `proposeNewTerms()`

## Events
The following events had the `deadline` parameter added to their signature:
- `NewTermsAccepted`
- `NewTermsProposed`

### CODE REVIEW

File: contracts/test/MapleLoan.t.sol [hunk 1]
```diff
@@ -263,11 +263,49 @@ contract MapleLoanTests is TestUtils {
         bytes[] memory data = new bytes[](1);
         data[0] = new bytes(0);
 
-        try loan.proposeNewTerms(mockRefinancer, data) { assertTrue(false, "Non-borrower was able to propose new terms"); } catch { }
+        try loan.proposeNewTerms(mockRefinancer, type(uint256).max, data) { assertTrue(false, "Non-borrower was able to propose new terms"); } catch { }
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:266]: Lets use `vm.expectRevert` for all of these assertions

File: contracts/test/MapleLoanInternals.t.sol [hunk 2]
```diff
@@ -2301,115 +2303,166 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
...
+        _loan.proposeNewTerms(address(_refinancer), deadline, calls);
 
         // Try to accept terms with a different refinancer than proposed.
         vm.expectRevert("MLI:ANT:COMMITMENT_MISMATCH");
-        _loan.acceptNewTerms(address(differentRefinancer), calls);
+        _loan.acceptNewTerms(address(differentRefinancer), deadline, calls);
+    }
+
+    function test_acceptNewTerms_commitmentMismatch_mismatchedDeadline() external {
+        Refinancer differentRefinancer = new Refinancer();
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:2393]: Should use the same refiancer for this test

File: contracts/test/MapleLoanInternals.t.sol [hunk 3]
```diff
@@ -2301,115 +2303,166 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
...
+        uint256 deadline = block.timestamp + 1;
+
         // Set _refinanceCommitment via _proposeNewTerms() using valid refinancer and empty calls array.
-        _loan.proposeNewTerms(address(_refinancer), calls);
+        _loan.proposeNewTerms(address(_refinancer), deadline, calls);
 
         // Try again with valid refinancer.
         vm.expectRevert("MLI:ANT:COMMITMENT_MISMATCH");
-        _loan.acceptNewTerms(address(_refinancer), calls);
+        _loan.acceptNewTerms(address(_refinancer), deadline, calls);
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:2353]: Lets add success calls after all of these

File: contracts/test/MapleLoanInternals.t.sol [hunk 4]
```diff
@@ -2421,16 +2474,57 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
...
+
+    MapleLoanInternalsHarness internal _loan;
+    Refinancer                internal _refinancer;
+
+    function setUp() external {
+        _loan       = new MapleLoanInternalsHarness();
+        _refinancer = new Refinancer();
+    }
+
+    function test_rejectNewTerms_commitmentMismatch() external {
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:2501]: Lets add cases for refinancer different address and different calls array

File: contracts/test/MapleLoan.t.sol [hunk 5]
```diff
@@ -438,9 +476,11 @@ contract MapleLoanTests is TestUtils {
 
         bytes[] memory calls = new bytes[](0);
 
-        loan.__setRefinanceCommitmentHash(keccak256(abi.encode(address(refinancer), calls)));
+        uint256 deadline = block.timestamp + 1;
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:479]: Same here lets just use deadline = block.timestamp

File: contracts/test/MapleLoan.t.sol [hunk 6]
```diff
@@ -456,7 +496,7 @@ contract MapleLoanTests is TestUtils {
         address refinancer = address(new EmptyContract());
         bytes[] memory data = new bytes[](1);
         data[0] = new bytes(0);
-        loan.__setRefinanceCommitmentHash(keccak256(abi.encode(refinancer, data)));
+        loan.__setRefinanceCommitmentHash(keccak256(abi.encode(refinancer, type(uint256).max, data)));
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:499]: Lets use deadline to be consistent with other tests

File: contracts/test/MapleLoan.t.sol [hunk 7]
```diff
@@ -465,11 +505,11 @@ contract MapleLoanTests is TestUtils {
         assertEq(loan.claimableFunds(),               0);
         assertEq(loan.drawableFunds(),                0);
 
-        try loan.acceptNewTerms(refinancer, data, 1) { assertTrue(false, "Able to accept terms"); } catch { }
+        try loan.acceptNewTerms(refinancer, type(uint256).max, data, 1) { assertTrue(false, "Able to accept terms"); } catch { }
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:508]: vm.expectRevert

File: contracts/test/Refinancer.t.sol [hunk 8]
```diff
@@ -92,16 +92,18 @@ contract RefinancerEndingPrincipalTest is BaseRefinanceTest {
...
-        interestRate_       = constrictToRange(interestRate_,       0,                MAX_RATE);
-        paymentInterval_    = constrictToRange(paymentInterval_,    1,                MAX_TIME);
-        paymentsRemaining_  = constrictToRange(paymentsRemaining_,  3,                MAX_PAYMENTS);
+        principalRequested_ = constrictToRange(principalRequested_, MIN_TOKEN_AMOUNT,    MAX_TOKEN_AMOUNT);  // Boundary increase so principal portion on 'paymentBreakdown' is always greater than 0
+        collateralRequired_ = constrictToRange(collateralRequired_, 0,                   MAX_TOKEN_AMOUNT);
+        gracePeriod_        = constrictToRange(gracePeriod_,        0,                   MAX_TIME);
+        interestRate_       = constrictToRange(interestRate_,       0,                   MAX_RATE);
+        paymentInterval_    = constrictToRange(paymentInterval_,    1,                   MAX_TIME);
+        paymentsRemaining_  = constrictToRange(paymentsRemaining_,  3,                   MAX_PAYMENTS);
+        deadline_           = constrictToRange(deadline_,           block.timestamp + 1, type(uint256).max);
```
Comment (lucas-manuel) [inline contracts/test/Refinancer.t.sol:106]: block.timestamp instead of block.timestamp + 1 everywhere

File: contracts/test/Refinancer.t.sol [hunk 9]
```diff
@@ -559,10 +582,10 @@ contract RefinancePrincipalRequestedTest is BaseRefinanceTest {
 
         bytes[] memory data = _encodeWithSignatureAndUint("increasePrincipal(uint256)", principalIncrease_);
 
-        loan.proposeNewTerms(address(refinancer), data);
+        loan.proposeNewTerms(address(refinancer), block.timestamp + 1, data);
```
Comment (lucas-manuel) [inline contracts/test/Refinancer.t.sol:585]: block.timestamp

File: contracts/test/Refinancer.t.sol [hunk 10]
```diff
@@ -749,6 +778,76 @@ contract RefinanceMultipleParameterTest is BaseRefinanceTest {
...
+        external
+    {
+        paymentInterval_    = constrictToRange(paymentInterval_,    1, MAX_TIME / 2);
+        newPaymentInterval_ = constrictToRange(newPaymentInterval_, 0, MAX_TIME);
+
+        setUpOngoingLoan(1 ether, 0, 0, 1, 1, paymentInterval_, 2);
+
+        assertEq(loan.paymentInterval(), paymentInterval_);
+
+        deadline_ = constrictToRange(deadline_, block.timestamp + 1, block.timestamp + MAX_TIME);
```
Comment (lucas-manuel) [inline contracts/test/Refinancer.t.sol:798]: block.timestamp

File: contracts/test/Refinancer.t.sol [hunk 11]
```diff
@@ -749,6 +778,76 @@ contract RefinanceMultipleParameterTest is BaseRefinanceTest {
...
+        
+        assertEq(loan.paymentInterval(), paymentInterval_);
+
+        bytes[] memory data = _encodeWithSignatureAndUint("setPaymentInterval(uint256)", newPaymentInterval_);
+
+        loan.proposeNewTerms(address(refinancer), deadline_, data);
+
+        emit log_named_uint("time     ", block.timestamp);
+        emit log_named_uint("deadline_", deadline_);
```
Comment (lucas-manuel) [inline contracts/test/Refinancer.t.sol:838]: Get rid of logs

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: contracts/MapleLoan.sol [hunk 12]
```diff
@@ -221,6 +222,12 @@ contract MapleLoan is IMapleLoan, MapleLoanInternals {
     /*** Miscellaneous Functions ***/
     /*******************************/
 
+    function rejectNewTerms(address refinancer_, uint256 deadline_, bytes[] calldata calls_) external override whenProtocolNotPaused {
+        require((msg.sender == _borrower) || (msg.sender == _lender), "L:RNT:NO_AUTH");
+
+        emit NewTermsRejected(_rejectNewTerms(refinancer_, deadline_, calls_), refinancer_, deadline_, calls_);
+    }
```
Comment (xhad) [inline contracts/MapleLoan.sol:229]: This is awesome thanks guys

File: contracts/test/MapleLoan.t.sol [hunk 13]
```diff
@@ -953,7 +1002,7 @@ contract MapleLoanTests is TestUtils {
         loan.__setFundsAsset(address(fundsAsset));
         loan.__setPrincipalRequested(amount);
         loan.__setPrincipal(amount);
-        loan.__setNextPaymentDueDate(block.timestamp + 1);
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:956]: Lets leave these as is as they are out of scope

File: contracts/test/MapleLoan.t.sol [hunk 14]
```diff
@@ -487,9 +534,11 @@ contract MapleLoanTests is TestUtils {
         loan.__setLender(address(this));
 
         address refinancer = address(new EmptyContract());
-        bytes[] memory data = new bytes[](1);
-        data[0] = new bytes(0);
-        loan.__setRefinanceCommitmentHash(keccak256(abi.encode(refinancer, data)));
+        uint256 deadline = block.timestamp;
```
Comment (lucas-manuel) [inline contracts/test/MapleLoan.t.sol:537]: All explicitly assigned deadlines should be `uint256 deadline = block.timestamp + 10 days`

File: contracts/test/MapleLoanInternals.t.sol [hunk 15]
```diff
@@ -1131,7 +1131,7 @@ contract MapleLoanInternals_RepossessTests is TestUtils {
     }
 
     function testFail_repossess_beforePaymentDue() external {
-        _loan.setNextPaymentDueDate(block.timestamp + 1);
+        _loan.setNextPaymentDueDate(block.timestamp);
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:1134]: Lets leavea all of these as is as they are out of scope

File: contracts/test/MapleLoanInternals.t.sol [hunk 16]
```diff
@@ -2279,137 +2280,146 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
         _loan.setCollateral(_defaultAmounts[0]);
     }
 
-    function test_acceptNewTerms_happyPath() external {
-        bytes[] memory calls = new bytes[](1);
+    function test_acceptNewTerms_commitmentMismatch_emptyCallsArray() external {
+        uint256 deadline = block.timestamp;
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:2284]: All explicitly assigned deadlines should be `uint256 deadline = block.timestamp + 10 days`

File: contracts/test/MapleLoanInternals.t.sol [hunk 17]
```diff
@@ -2419,18 +2429,128 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
...
+        _loan.proposeNewTerms(address(_refinancer), deadline, calls);
+
+        vm.expectRevert("MLI:RNT:COMMITMENT_MISMATCH");
+        _loan.rejectNewTerms(address(_refinancer), deadline + 1, calls);
+
+        _loan.rejectNewTerms(address(_refinancer), deadline, calls);
+    }
+
+    function test_rejectNewTerms_commitmentMismatch_mismatchedCalls() external {
+        Refinancer anotherRefinancer = new Refinancer();
```
Comment (lucas-manuel) [inline contracts/test/MapleLoanInternals.t.sol:2523]: Should not use `anotherRefinancer` at all in this test

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Pull request merged**